File: backend/app/services/tier_classifier.py
"""
Tier Classifier — Signal-based classification of entitlements by business criticality / audit risk.

Tiers:
- critical (level 1): Admin/full access to production, PII/financial data
- high (level 2): Write access to production, sensitive data, cross-system
- medium (level 3): Read access to production, write to non-prod
- low (level 4): Read-only non-sensitive, internal tools
"""

from app.models.entitlement import Entitlement

CRITICAL_KEYWORDS = [
    "production", "prod", "pii", "financial", "payment", "credential",
    "secret", "root", "infrastructure", "superuser", "master", "admin-all",
    "customer-data", "healthcare", "hipaa", "pci", "sox", "gdpr",
    "encryption-key", "signing-key", "private-key",
]

HIGH_KEYWORDS = [
    "staging", "sensitive", "internal-api", "service-account",
    "cross-system", "database", "data-warehouse", "analytics",
    "monitoring", "logging", "ci-cd", "pipeline", "deploy",
    "container-registry", "artifact", "backup",
]

BROAD_SCOPE_KEYWORDS = [
    "all", "unlimited", "unrestricted", "wildcard", "*",
    "everything", "full-scope", "global",
]

ACCESS_WEIGHTS = {
    "admin": 40,
    "full": 40,
    "superadmin": 45,
    "write": 25,
    "read-write": 30,
    "execute": 20,
    "read": 10,
    "view": 5,
    "list": 5,
}


def classify_risk_tier(
    entitlement: Entitlement,
    llm_risk_assessment: str | None = None,
) -> str:
    """
    Classify an entitlement into a risk tier based on signals.

    Returns: "critical", "high", "medium", or "low"
    """
    score = 0

    # 1. Access level signals
    access_level = entitlement.access_level.lower().strip()
    score += ACCESS_WEIGHTS.get(access_level, 15)

    # 2. Resource sensitivity signals
    combined_text = (
        f"{entitlement.description} {entitlement.resource_name} "
        f"{entitlement.resource_type} {entitlement.name}"
    ).lower()

    has_critical = any(kw in combined_text for kw in CRITICAL_KEYWORDS)
    has_high = any(kw in combined_text for kw in HIGH_KEYWORDS)

    if has_critical:
        score += 30
    elif has_high:
        score += 20

    # 3. Broad scope signals
    has_broad_scope = any(kw in combined_text for kw in BROAD_SCOPE_KEYWORDS)
    if has_broad_scope:
        score += 15

    # 4. Missing safeguards escalation
    if not entitlement.conditions or not entitlement.conditions.strip():
        score += 10
    if (
        not entitlement.business_justification
        or not entitlement.business_justification.strip()
    ):
        score += 5

    # 5. LLM risk assessment override (if available)
    if llm_risk_assessment:
        llm_tier = llm_risk_assessment.lower().strip()
        if llm_tier == "critical":
            score = max(score, 70)
        elif llm_tier == "high":
            score = max(score, 45)

    # 6. Tier mapping
    if score >= 60:
        return "critical"
    elif score >= 40:
        return "high"
    elif score >= 20:
        return "medium"
    else:
        return "low"


def get_tier_metadata(tier: str) -> dict:
    """Get metadata for a given tier."""
    metadata = {
        "critical": {
            "level": 1,
            "color": "#DC2626",
            "review_frequency": "monthly",
            "approval_required": True,
            "description": (
                "Admin/full access to production systems, PII/financial data access, "
                "infrastructure-level permissions. Requires monthly recertification."
            ),
        },
        "high": {
            "level": 2,
            "color": "#F59E0B",
            "review_frequency": "quarterly",
            "approval_required": True,
            "description": (
                "Write access to production, access to sensitive data, "
                "cross-system service accounts. Requires quarterly recertification."
            ),
        },
        "medium": {
            "level": 3,
            "color": "#3B82F6",
            "review_frequency": "semi-annually",
            "approval_required": False,
            "description": (
                "Read access to production, write access to non-prod, "
                "standard business application access."
            ),
        },
        "low": {
            "level": 4,
            "color": "#10B981",
            "review_frequency": "annually",
            "approval_required": False,
            "description": (
                "Read-only access to non-sensitive systems, documentation wikis, "
                "internal tools. Annual self-certification."
            ),
        },
    }
    return metadata.get(tier, metadata["medium"])

File: frontend/src/lib/tier-classifier.ts
// Client-side tier classifier — mirrors backend/app/services/tier_classifier.py

const CRITICAL_KEYWORDS = [
  "production", "pii", "financial", "payment", "credential", "secret", "root",
  "admin-all", "customer-data", "healthcare", "hipaa", "pci", "sox", "gdpr",
  "encryption-key",
];

const HIGH_KEYWORDS = [
  "staging", "sensitive", "internal-api", "service-account", "cross-system",
  "database", "data-warehouse", "analytics", "monitoring", "logging", "ci-cd",
  "deploy", "container-registry",
];

const BROAD_SCOPE_KEYWORDS = [
  "all", "unlimited", "unrestricted", "*", "everything", "full-scope", "global",
];

export interface TierResult {
  tier: string;
  tier_level: number;
  color: string;
  review_frequency: string;
  approval_required: boolean;
}

const TIER_META: Record<string, Omit<TierResult, "tier">> = {
  critical: { tier_level: 1, color: "#DC2626", review_frequency: "monthly", approval_required: true },
  high: { tier_level: 2, color: "#F59E0B", review_frequency: "quarterly", approval_required: true },
  medium: { tier_level: 3, color: "#3B82F6", review_frequency: "semi-annual", approval_required: false },
  low: { tier_level: 4, color: "#10B981", review_frequency: "annual", approval_required: false },
};

export function classifyTier(data: {
  access_level: string;
  resource_name: string;
  resource_type: string;
  description: string;
  conditions?: string | null;
  business_justification?: string | null;
}): TierResult {
  let score = 0;
  const searchText = [
    data.access_level, data.resource_name, data.resource_type,
    data.description, data.conditions, data.business_justification,
  ].filter(Boolean).join(" ").toLowerCase();

  // Access level scoring
  const al = data.access_level.toLowerCase();
  if (["superadmin", "admin", "full"].some((k) => al.includes(k))) score += 40;
  else if (al.includes("read-write")) score += 30;
  else if (al.includes("execute")) score += 25;
  else if (al.includes("write")) score += 20;
  else if (al.includes("list")) score += 10;
  else if (["read", "view"].some((k) => al.includes(k))) score += 5;

  // Resource sensitivity
  for (const kw of CRITICAL_KEYWORDS) {
    if (searchText.includes(kw)) { score += 30; break; }
  }
  for (const kw of HIGH_KEYWORDS) {
    if (searchText.includes(kw)) { score += 20; break; }
  }

  // Broad scope
  for (const kw of BROAD_SCOPE_KEYWORDS) {
    if (searchText.includes(kw)) { score += 15; break; }
  }

  // Missing safeguards
  if (!data.conditions) score += 10;
  if (!data.business_justification) score += 5;

  // Map to tier
  let tier: string;
  if (score >= 60) tier = "critical";
  else if (score >= 40) tier = "high";
  else if (score >= 20) tier = "medium";
  else tier = "low";

  return { tier, ...TIER_META[tier] };
}
